### ISSUE TITLE
Add "Accepted" and "Queue Position" fields to signups

### ISSUE THREAD

**Issue by fubu** [amiv-eth/amivapi#71]
I think the discussion happened on the IT dinner, but it never made it into the issue tracker, so here it is:

We should consider supporting other signup modes apart from the (default) FCFS. There have been vivid discussions at AMIV in the past few years about more fairness for signups, for example:
- Fair-FCFS: Signups are only confirmed X minutes after the registration opened. Before confirming, the signup order is shuffled. This basically gives everybody a X-minute window to sign up with equal chances for everybody in that time frame. Afterwards it is normal FCFS.
- Hand-Picking: Signups must be confirmed by hand (e.g. through the event admin).
- Constraint-Based: Some spots are only given out under specific constraints, e.g. keep 10 spots open for women/MAVTies/...

Especially the "Fair-FCFS" got a lot of upvotes in the past, and "Hand-Picking" can be useful every now and then. "Constraint-Based" could get rather complex, so we should maybe focus on the other two. Implications for the API:
- Signups need some kind of sorting built in.
- A signup needs another flag to mark the accepted ones.
- An event needs a field to specify the "accept mode".
- Of course some logic to accept signups based on the chosen mode.

(Please note that the "accepted" flag is probably required anyways when implementing #48)

What do you think?

**Comment by NotSpecial** [comment]
I am not sure if the API needs to do this.
An "accepted" flag is necessary, yes.

But instead of bulding several automatizations into the api if could just be "hand-picking" by default and any client could do the rest.

Then a client could implement different strategies.

Pro: More flexibility, switching to a new mechanism doesn't require API change
Con: No automatization (or at least more complicated)

**Comment by hermannsblum** [comment]
I support Alex solution.
The event-admin can use/write a tool which picks the accepted signups. Therefore we have the biggest possible flexibility in the API.

But, However, we really need an 'accepted' flag

**Comment by fubu** [comment]
If clients need to implement such strategies, then *all* clients need to do that. Otherwise, we're back to square one: I'll just write my own client then, which only sends "accepted" signups. (Ah, let's restrict sending "accept" flags to certain clients then -- an we're back to unneccessary complexity)

Also, hand picking alone doesn't solve the fairness problem. If it boils down to the event admin "randomly" picking from the signups, we're no further than now (people will complain that the admins prefer their own friends -- and the peer pressure will be there, believe me). I believe that at least some basic functionality besides the simple "FCFS" mode should be handled by the API.

I'm also a big fan of instant feedback: If I sign up, I want to be certain that I have gotten a slot, and not wait for the mercy of the event admin to make his selection. Especially not for the many default cases (the plain old FCFS works quite well most of the time).

So, I'm reformulating my proposal:

1) The API should require signups to be accepted (with the flag we all agree on). By default, events are the traditional FCFS (which auto-accepts in the order the signups were POSTed). Events can also be configured to require hand-picking (this still only uses the `accepted` flag, except that here it must be set explicitly by an event admin). I think this is the part we already agree.

2) As a third possibility, events can be configured to use the "fair-fcfs" algorithm as mentioned above. This would require some mechanism to order signups (besides `_created`). IMHO it would also make sense to have this mode as the default - if there is less signups than spots in the initial randomization phase, the outcome is the same as with plain old FCFS. If there's more, this one is fairer.

By the way, when skipping the "constraint-based" algorithm above (I don't need that), I personally don't see any other sensible algorithm we would have to implement in the future. What is more: For clients, the only thing that matters is that a signup might not directly be accepted (they have to consider this anyways). They don't have to send any other data. The receive the same data for all algorithms. Only deep inside the backend there might be a change. And this is exactly what the API was built for: to decouple frontend from our data processing :)

**Comment by hermannsblum** [comment]
1) the amiss board decided a year ago to default-use fair-FCFS (but got nobody implementing it) and i totally agree that there is no reason at all to use FCFS.
As this turns it down to 2 different selection policies, I agree that implementing this into the api is worth a thought.

But how does it look exactly?
a) a process outside of eve handles the accepted-flag and will execute if it is 5 min after a signup-window-start
b) every new signup will check and eventually change the accepted-flags of signups made in the first 5 min (inside of a hook)
c) any ideas?

Also, we have a feedback problem. Users won't get feedback before 5 min after the signup-window-start. This can be dealt with in different ways
a) we output a message that they have to look later (how do we do this in the best REST-way?) <- I prefer this
b) we send an email/notification/... after the 5min period
c) Live-view of the accepted-flags

For a coexistance of Fair-FCFS and hand-picking:
* What is the best way to define this setup in the event? 
    * Different spot-fields for both policies
    * Default Fair-FCFS, additional spots for hand-picking in a non-required field <- my favorite
    * only one policy for one event
* How does the handpicking work?
    * If the e.g. helper already signed up, how can the admin indicate that this is a handpicked person and the FFCFS-algorithm can pick another person for the FFCFS-places?
    * Maybe an additional 'helper'-field?  In this case Helping will always work (no field in the event required) and it is just in the hands of the event organizer how many persons he wants to 'hand-pick', but he still can manage all attending people via the api

**Comment by fubu** [comment]
Fair point about dropping normal FCFS entirely in favor of Fair-FCFS :+1: 

> But how does it look exactly?

I would run a background task X minutes after the signup period started. (I think an async background worker queue would make sense for some other stuff as well, e.g. rewriting the `.forward` files. Maybe `cron` is good enough for that; also, [Celery](https://github.com/celery/celery) is a nice tool for that job)
After the initial grace period, signups could accept themselves automatically if there are spots left. Or re-use the previous async task to avoid race-conditions (two concurrent signups when only one spot is left). The latter also simplifies handling of signup cancellations (just re-run the async task) and especially designating helpers from normal signups (which are kind-of cancellations).

> Also, we have a feedback problem. Users won't get feedback before 5 min after the signup-window-start.

People will get used to waiting on the acceptance mail -- IMHO the initial "fair" period should be rather short, e.g. the mentioned 5 minutes (but not "1 hour"). Maybe we should not even allow to customize this per event, but rather have a global default value for that (which can be tweaked if we ever find out that it's too short/long). The "fairness" part is mostly about randomizing the success rate of bots/scripts and other environmental influences in the first few minutes, not about giving the same chances to people signing up an hour later.

As for the RESTy response, there is the status code [**202 Accepted**](https://tools.ietf.org/html/rfc7231#section-6.3.3):

> The 202 (Accepted) status code indicates that the request has been
> accepted for processing, but the processing has not been completed.
> The request might or might not eventually be acted upon, as it might
> be disallowed when processing actually takes place.  There is no
> facility in HTTP for re-sending a status code from an asynchronous
> operation.
>
> The 202 response is intentionally noncommittal.  Its purpose is to
> allow a server to accept a request for some other process (perhaps a
> batch-oriented process that is only run once per day) without
> requiring that the user agent's connection to the server persist
> until the process is completed.  The representation sent with this
> response ought to describe the request's current status and point to
> (or embed) a status monitor that can provide the user with an
> estimate of when the request will be fulfilled.

I'm not sure if this response code really fits (after all, the signup was processed, just not yet accepted -- and some never will), and IMHO a **201 Created** (as it is now) is also a reasonable choice.

Nonetheless, a subsequent GET request for the newly created signup should contain the `accepted` flag with a value of `null` (= no decision has been made yet) or `true/false` (= after the background job / event admin has selected the signups).

Example flow:

```
CLIENT            API

POST     ----->   /events 
         <-----   202 Accepted
                  (Location: /events/<id>)

GET      ----->   /events/<id>
         <-----   202 Accepted
                  (accepted = null)

      (time passes)

GET      ----->   /events/<id>
         <-----   200 OK
                  (accepted = true|false)
```

> For a coexistance of Fair-FCFS and hand-picking:
> [...]
>
> * Maybe an additional 'helper'-field? In this case Helping will always work (no field in the event required) and it is just in the hands of the event organizer how many persons he wants to 'hand-pick', but he still can manage all attending people via the api

From what I understand, "co-existence" means the need to have some spots reserved for "helpers". If that's the only use-case, I would go with the quoted solution. I still think it could make sense to configure an event to only accept hand-picked signups, so we should have two options there: "Fair-FCFS" or "hand-picked", both with the additional "helper" add-on.

On a side note: Of course an event organizer could add unlimited helpers just for fun, but I think that's a social problem which should be handled by supervision (Kulturvorstand) -- assuming that managing the signups for one event is only possible for the event owner or users with the "event admin" role.

When adding such a helper flag, the `signup_count` attribute from #48 should only count non-helpers, and `signups` should only contain non-helper signups. Additionally we should consider adding additional admin-only fields `helpers` (only the helper signups) and `helper_count`.

**Comment by hermannsblum** [comment]
>I still think it could make sense to configure an event to only accept hand-picked signups, so we should have two options there: "Fair-FCFS" or "hand-picked", both with the additional "helper" add-on.

To sum it up for this suggestion:
* a new "accepted" and a new "helper" field for signups
* a new enumeration "Fair-FCFS" and "handpicking" in the field "selection policy" of the event
* a new field "helper_count" in the event only visible for event-owner or event-admin

A background process will handle the fair-fcfs and our permission-system will need to take care of the Handpicking as well as marking people as helper. Both will be done by event-admin or event-owner.

Everyone OK with this?

**Comment by NotSpecial** [comment]
We have discussed this once again and have settled on this:

* a "accepted" field on signups, will be Null first and will be set according to selection policy
* for selection policy, we will include FFCFS with a (short) waiting time or "Handpicking" (together with appropiate field like hermann wrote)
* a new field "position" for every signup that orders the signups. After the FFCFS window, those places will be shuffled. The first n positions for n spots will be selected. After the FFCFS window, they will just count up normally

So far as you stated. But we have decided against a field for helpers, since its usefulness in itself is not that great while requiring a lot of maintenance:

*all helpers need to sign up
*only few events actually want "random" helpers (ski weekend, sushi are the only examples right now)
*helpers need to be organised with different tools anyway (e.g. google doc of helper shifts)
*creates new problematic situations:
**What if to many helpers sign up
**What if someone wanted to be a helper, therefore someone else gets his spot and then he cannot be a helper anymore for some reason

**Comment by NotSpecial** [comment]
I have split up the issues. This one is only about adding the necessary fields.

Fair FCFS is low priority since it is a defence against a thread that might never be really problematic.
(Script signup to gain a (probably small) headstart against one click signup on website)

But since it might indeed become an issue it has now its own issue #110

**Comment by fubu** [comment]
Just for completeness (I'm okay with postponing this to later):

> Fair FCFS is low priority since it is a defence against a thread that might never be really problematic.
(Script signup to gain a (probably small) headstart against one click signup on website)

There is another motivation behind Fair-FCFS besides the "scripted signups": Even without bots there is a lot of variables involved with the order of the signups (network/typing speed, slow/laggy internet connection, lecture which lasted 5mins too long). Right now it's "I have to be online exactly at this time, 30seconds later and I can just skip it anyways"; Fair-FCFS takes away a lot of the stress involved with signing up. I think it's still worth it to pursue the idea, as I think it will also lead to more people signing up to events (giving us a more realistic picture how many people really were interested in the event) - right now, if you're 30 seconds too late, I'm afraid many people will just not sign up anymore, as "there is no chance to get in anyways".
(Just to reiterate: We're not talking about a 30 minute window, more of a 1-5 minute window; and after this grace period the remaining spots are traditional FCFS again.)

**Comment by NotSpecial** [comment]
Thats a very valid point. Once the more pressing issues are resolved we should really take a look at this again.

**Comment by Tar-Tarus** [comment]
Should the "QUEUE Position " Field in the eventsignup appear only for people who have not been accepted or should it be available for all signups?

**Comment by fubu** [comment]
The "queue position" is technically only necessary to order signups under the Fair-FCFS scheme, in which *all signups* need to have a position. (These positions should monotonically increase and at least loosely follow the `_created` time; under Fair-FCFS, they additionally get shuffled after the timeout.)

**Comment by Tar-Tarus** [comment]
The signups now have a field *'position'* added after a GET Hook on __*/eventsignups*__ . They are asigned by querying the DB after signups on the same event with a *_created* tag that is older than that of the requested item. Thus we get the position of the requested signup in the signups list of a particular event. This position also changes according to the changes in signups to that event, meaning that if a participant is deleted or unregisters to that particular event, then on the next GET, the position information will be updated. Hope this fits with the idea^^

**Comment by Tar-Tarus** [comment]
Since the __*position*__ field has been implemented and tested, and since we are using a FCFS system (at least until we have a final decision on implementing Fair FCFS with shuffle), we do not need an __*accepted*__ which can simply be replaced by a comparison of *position* with *spots*. Thus this issue can be closed and further discussion/development relates to #110 .

**Comment by fubu** [comment]
In my opinion even with FCFS only the aditional `accepted` field is better than deriving the acceptance to an event by only comparing the `EventSignup.position` with `Event.spots`:

1. It is crystal clear by looking at the signup itself. In general we aim(ed?) to put such information explicitly into a resource instead of letting the client derive anything from other fields (especially from other resources), especially if the logic of deriving the property should remain "private" (= opaque to the client). As a counter-example (see #48), I completely agree to let the client handle the question "how many open spots are left for a given event" instead of having an `remaining_spots` property, since I see no reason to keep the actual calculation logic behind that property private, as I don't see any possible extension which would warrant that (we already clearly dismissed the helper system with seperate spot counting behaviour).
2. It makes it easier to introduce additional signup modes like "hand-picked" or "Fair-FCFS" in the future without breaking the API, as you already pointed out yourself.
3. Hypothetically you could implement the `position` using a global counter (over all events; kind of like `_created` inherently works), maybe even as immutable (assign-once, not re-assigning after a signup is deleted), without breaking the basic assumption of the field: A higher `position` number means "lower in importance for calculating/ordering the signup list" (as described above, without the `position` field the `_created` field would've delivered the same semantics, but only in the general FCFS case).

As a result, adding the field does not make the API more complicated, but enables us to optimize/enhance signup functionality independent of the signup acceptance state.

(By the way, the `accepted` flag should probably never automatically change from "true" to "false", except on manual intervention by an event admin or the signup owner.)

PS: I think for additional signup modes like "hand-picking" you'd probably also need a new enumerated field on `Events`; it wouldn't just work with the `accepted` flag alone.

**Comment by NotSpecial** [comment]
This seems to be implemented since 52732bddc0386dc0ab9f2b406c6505d41209e38a :)
